Trabajaremos con el archivo: **Life_Expectancy_Data**

In [1]:
#Importamos las librerías
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df = pd.read_csv("C:/Users/Layla/Desktop/Docencia/Coder House/Materiales complementarios/Mod 6 - Data Science/3. Mejora de Modelos de Machine Learning II/Datos/Life_Expectancy_Data.csv", sep = ",")
df.head()

Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   ...  \
0             62     0.01               71.279624         65.0      1154  ...   
1             64     0.01               73.523582         62.0       492  ...   
2             66     0.01               73.219243         64.0       430  ...   
3             69     0.01               78.184215         67.0      2787  ...   
4             71     0.01                7.097109         68.0      3013  ...   

   Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0    6.0               8.16         65.0        0.1  584.259210  33736494.0   
1   58.0               8.18         62.0        0.1  612.696514    327582.0   
2   62.0               8.13         64.0        0.1  631.744976  31731688.0   
3   67.0               8.52         67.0        0.1  669.959000   3696958.0   
4   68.0               7.87         68.0        0.1   63.537231   2978599.0   

    thinness  1-19 years   thinness 5-9 years  \
0                   17.2                 17.3   
1                   17.5                 17.5   
2                   17.7                 17.7   
3                   17.9                 18.0   
4                   18.2                 18.2   

   Income composition of resources  Schooling  
0                            0.479       10.1  
1                            0.476       10.0  
2                            0.470        9.9  
3                            0.463        9.8  
4                            0.454        9.5  

[5 rows x 22 columns]

In [2]:
#Dimension del dataset
print(df.shape)

(2938, 22)


In [3]:
#Tipo de dato para cada columna
df.dtypes

Country                             object
Year                                 int64
Status                              object
Life expectancy                    float64
Adult Mortality                    float64
infant deaths                        int64
Alcohol                            float64
percentage expenditure             float64
Hepatitis B                        float64
Measles                              int64
 BMI                               float64
under-five deaths                    int64
Polio                              float64
Total expenditure                  float64
Diphtheria                         float64
 HIV/AIDS                          float64
GDP                                float64
Population                         float64
 thinness  1-19 years              float64
 thinness 5-9 years                float64
Income composition of resources    float64
Schooling                          float64
dtype: object

In [4]:
#Información del dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10   BMI                             2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

In [5]:
#Vemos que columnas tenemos!
df.columns

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

In [6]:
#Renombramos los nombres de las columnas que poseen espacios
df.rename(columns={'Life expectancy ':'Life_expectancy','Adult Mortality':'Adult_Mortality','infant deaths': 'infant_deaths','percentage expenditure':'percentage_expenditure','Hepatitis B':'Hepatitis_B','Measles ':'Measles',' BMI ':'BMI', 'under-five deaths ':'under_five_deaths', 'Total expenditure':'Total_expenditure', 'Diphtheria ': 'Diphtheria',' HIV/AIDS':'HIV/AIDS', ' thinness  1-19 years':'thinness_1_19_years',' thinness 5-9 years':'thinness_5_9_years','Income composition of resources':'Income_composition_of_resources'}, inplace=True)

In [7]:
#Verificamos que se hayan realizado los cambios!
df.columns

Index(['Country', 'Year', 'Status', 'Life_expectancy', 'Adult_Mortality',
       'infant_deaths', 'Alcohol', 'percentage_expenditure', 'Hepatitis_B',
       'Measles', 'BMI', 'under_five_deaths', 'Polio', 'Total_expenditure',
       'Diphtheria', 'HIV/AIDS', 'GDP', 'Population', 'thinness_1_19_years',
       'thinness_5_9_years', 'Income_composition_of_resources', 'Schooling'],
      dtype='object')

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life_expectancy                  2928 non-null   float64
 4   Adult_Mortality                  2928 non-null   float64
 5   infant_deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage_expenditure           2938 non-null   float64
 8   Hepatitis_B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10  BMI                              2904 non-null   float64
 11  under_five_deaths                2938 non-null   int64  
 12  Polio               

In [9]:
#Eliminacion de valores nulos
df = df.dropna()

In [10]:
#Separacion en X e y
y = df['Life_expectancy'].to_numpy() 

for n in df.columns:
  if str(df[n].dtype) == 'object' or str(df[n].dtype) == 'category':
    df[n] = df[n].astype('category').cat.codes
X = df.drop(['Life_expectancy'], axis=1).to_numpy()
df.head()

Country  Year  Status  Life_expectancy  Adult_Mortality  infant_deaths  \
0        0  2015       1             65.0            263.0             62   
1        0  2014       1             59.9            271.0             64   
2        0  2013       1             59.9            268.0             66   
3        0  2012       1             59.5            272.0             69   
4        0  2011       1             59.2            275.0             71   

   Alcohol  percentage_expenditure  Hepatitis_B  Measles  ...  Polio  \
0     0.01               71.279624         65.0     1154  ...    6.0   
1     0.01               73.523582         62.0      492  ...   58.0   
2     0.01               73.219243         64.0      430  ...   62.0   
3     0.01               78.184215         67.0     2787  ...   67.0   
4     0.01                7.097109         68.0     3013  ...   68.0   

   Total_expenditure  Diphtheria  HIV/AIDS         GDP  Population  \
0               8.16        65.0       0.1  584.259210  33736494.0   
1               8.18        62.0       0.1  612.696514    327582.0   
2               8.13        64.0       0.1  631.744976  31731688.0   
3               8.52        67.0       0.1  669.959000   3696958.0   
4               7.87        68.0       0.1   63.537231   2978599.0   

   thinness_1_19_years  thinness_5_9_years  Income_composition_of_resources  \
0                 17.2                17.3                            0.479   
1                 17.5                17.5                            0.476   
2                 17.7                17.7                            0.470   
3                 17.9                18.0                            0.463   
4                 18.2                18.2                            0.454   

   Schooling  
0       10.1  
1       10.0  
2        9.9  
3        9.8  
4        9.5  

[5 rows x 22 columns]

Iniciaremos aplicando el método de **KFold**. 

In [11]:
##KFold - Train y Test
from sklearn.model_selection import KFold
KFold = KFold(n_splits=5, random_state=0, shuffle=True)
train, test = list(KFold.split(X))[0]
X_train = X[train]
X_test = X[test]
y_train = y[train]
y_test = y[test]

In [12]:
#Entrenar arbol y encontrar el mejor alpha con GridSearchCV y KFold
from sklearn.model_selection import KFold
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import GridSearchCV
t = DecisionTreeRegressor()
par = list(np.arange(0.0, 1., step=0.05))
cv = KFold(n_splits=5, random_state=0, shuffle=True)
tunner = GridSearchCV(estimator=t, param_grid={'ccp_alpha':par}, cv=cv)
_=tunner.fit(X_train, y_train)

In [13]:
print('Train score KFold Tree: ' + str(tunner.score(X_train, y_train)))
print('Test score KFold Tree: ' + str(tunner.score(X_test, y_test)))

Train score KFold Tree: 0.9532587993823117
Test score KFold Tree: 0.9152860839590284


**Interpretaciones**

Al analizar las métricas obtenidas, podemos identificar que los valores en Train y Test, son parecidos y significativamente altos. A continuación se detallan los resultados alcanzados:

* Train score KFold Tree: 0.9532587993823117
*Test score KFold Tree: 0.9152860839590284


In [39]:
#StratifiedKFold - Train y Test
from sklearn.model_selection import StratifiedKFold
skf = StratifiedKFold(n_splits=5, random_state=0, shuffle=True)
train_2, test_2 = list(skf.split(X, y))[0]
X_train_2 = X[train_2]
X_test_2 = X[test_2]
y_train_2 = y[train_2]
y_test_2= y[test_2]

ValueError: Supported target types are: ('binary', 'multiclass'). Got 'continuous' instead.

Como se puede observar del error obtenido, no podemos utilizar StratifiedKFold dado que los tipos de destino admitidos son: 'binario' o 'multiclase'. En cambio, nuestra variable target: **Life_expectancy** es de tipo 'continua'.

**¿Qué más podríamos hacer?** Es posible también aplicar algún otro algoritmo de Regresión para verificar la performance del modelo. 